File: src/quantized/routes/reflectivity.py
"""Thin specular-reflectivity routes.

Wraps the finished W3 calc helpers (``calc.reflectivity.parratt_refl`` — golden vs
MATLAB parrattRefl — and ``calc.sld`` SLD profile / presets). The route builds the
Q grid, validates the layer stack, calls the pure functions, and serializes. No
physics here; the recursion + Névot-Croce roughness live in ``calc/``.
"""

from __future__ import annotations

from typing import Annotated, Any, Literal

import numpy as np
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from quantized.calc.refl_fit import fit_reflectivity
from quantized.calc.reflectivity import parratt_refl
from quantized.calc.sld import refl_sld_presets, sld_profile
from quantized.routes._errors import call_calc
from quantized.routes._payload import to_jsonable

router = APIRouter(prefix="/api/reflectivity", tags=["reflectivity"])

# A layer row is [thickness Å, SLD_real Å⁻², SLD_imag Å⁻², roughness Å].
Layer = list[float]


class SimulateRequest(BaseModel):
    """Simulate R(Q) from a layer stack over a linear Q grid."""

    layers: list[Layer] = Field(min_length=2)
    q_min: float = Field(default=0.005, gt=0.0)
    q_max: float = Field(default=0.25, gt=0.0)
    n_points: int = Field(default=400, ge=2, le=20000)
    roughness: bool = True
    scale: float = 1.0
    background: float = 0.0
    resolution: float | None = None  # dQ/Q (constant relative resolution)


class SldProfileRequest(BaseModel):
    """SLD(z) depth profile from a layer stack (error-function interfaces)."""

    layers: list[Layer] = Field(min_length=2)
    n_points: int = Field(default=500, ge=2, le=20000)
    padding: float = Field(default=50.0, ge=0.0)


def _validate_layers(layers: list[Layer]) -> None:
    if any(len(row) != 4 for row in layers):
        raise HTTPException(
            status_code=422,
            detail="each layer must be [thickness, sld_real, sld_imag, roughness]",
        )


@router.get("/presets")
def get_presets() -> dict[str, Any]:
    """Material SLD presets (name/formula/sldX/sldN/sldImag/density)."""
    return {"presets": to_jsonable(refl_sld_presets())}


@router.post("/simulate")
def simulate(req: SimulateRequest) -> dict[str, Any]:
    """Specular reflectivity R(Q) for the layer stack over [q_min, q_max]."""
    _validate_layers(req.layers)
    if req.q_max <= req.q_min:
        raise HTTPException(status_code=422, detail="q_max must exceed q_min")
    q = np.linspace(req.q_min, req.q_max, req.n_points)
    r = call_calc(parratt_refl,
        q,
        req.layers,
        roughness=req.roughness,
        scale=req.scale,
        background=req.background,
        resolution=req.resolution,
    )
    return {"q": to_jsonable(q), "r": to_jsonable(r)}


@router.post("/sld-profile")
def sld_profile_route(req: SldProfileRequest) -> dict[str, Any]:
    """SLD(z) depth profile (error-function interfaces) for the layer stack."""
    _validate_layers(req.layers)
    z, sld = call_calc(sld_profile, req.layers, n_points=req.n_points, padding=req.padding)
    return {"z": to_jsonable(z), "sld": to_jsonable(sld)}


# ── fit to measured data (audit P2.2) ────────────────────────────────────────

# A fit evaluates the model ~10-100 times per free parameter, and a resolution-
# smeared evaluation costs 21x a plain one; these caps keep one request to
# seconds, not minutes.
FIT_MAX_POINTS = 20_000
FIT_MAX_PARAMETERS = 200
FIT_MAX_CHANNELS = 4

FiniteFloat = Annotated[float, Field(allow_inf_nan=False)]


class FitParameter(BaseModel):
    """One model parameter: ``L{i}.{thickness|sld|isld|roughness|msld}``,
    ``scale``, ``background`` or a per-channel scale/background name."""

    name: str = Field(min_length=1, max_length=64)
    value: FiniteFloat
    vary: bool = False
    min: FiniteFloat | None = None
    max: FiniteFloat | None = None
    tie: str | None = Field(default=None, max_length=64)


class FitChannel(BaseModel):
    """One measured curve. ``dq`` is a per-point 1-sigma resolution unless
    ``dq_is_fwhm``; ``resolution`` is a constant dQ/Q used when ``dq`` is absent."""

    q: list[FiniteFloat] = Field(min_length=2, max_length=FIT_MAX_POINTS)
    r: list[FiniteFloat] = Field(min_length=2, max_length=FIT_MAX_POINTS)
    dr: list[FiniteFloat] | None = Field(default=None, max_length=FIT_MAX_POINTS)
    dq: list[FiniteFloat] | None = Field(default=None, max_length=FIT_MAX_POINTS)
    dq_is_fwhm: bool = False
    resolution: float | None = Field(default=None, ge=0.0, le=0.5, allow_inf_nan=False)
    spin: Literal["+", "-"] | None = None
    q_min: FiniteFloat | None = None
    q_max: FiniteFloat | None = None
    scale: str = Field(default="scale", max_length=64)
    background: str = Field(default="background", max_length=64)
    label: str | None = Field(default=None, max_length=120)


class FitRequest(BaseModel):
    parameters: list[FitParameter] = Field(min_length=1, max_length=FIT_MAX_PARAMETERS)
    channels: list[FitChannel] = Field(min_length=1, max_length=FIT_MAX_CHANNELS)
    weighting: Literal["dr", "log"] = "dr"
    max_nfev: int = Field(default=2000, ge=1, le=20_000)


@router.post("/fit")
def fit_route(req: FitRequest) -> dict[str, Any]:
    """Fit the layer model to one or more measured reflectivity curves."""
    channels = []
    for ch in req.channels:
        c = ch.model_dump()
        for key in ("q_min", "q_max"):
            if c[key] is None:
                del c[key]
        channels.append(c)
    out = call_calc(
        fit_reflectivity,
        [p.model_dump() for p in req.parameters],
        channels,
        weighting=req.weighting,
        max_nfev=req.max_nfev,
    )
    result: dict[str, Any] = to_jsonable(out)
    return result

File: frontend/src/lib/api/reflectivity.ts
// /api/reflectivity/* wrappers — split out of lib/api.ts (R8 bundle-diet
// pass, 2026-08-23; see api/reference.ts's header for why): its one
// consumer is the lazy reflectivity workshop. NOT re-exported by
// lib/api.ts; useReflectivity.ts imports directly from this path.

import { getJSON, postJSON } from "./http";
import type { SldPreset } from "../types";
import type { components } from "./schema";

/** Material SLD presets for building reflectivity models. */
export function reflPresets(): Promise<{ presets: SldPreset[] }> {
  return getJSON("/api/reflectivity/presets");
}

/** A layer row: [thickness Å, SLD_real Å⁻², SLD_imag Å⁻², roughness Å]. */
export type ReflLayer = [number, number, number, number];

/** Simulate specular reflectivity R(Q) from a layer stack (Parratt recursion). */
export function reflSimulate(body: {
  layers: ReflLayer[];
  q_min?: number;
  q_max?: number;
  n_points?: number;
  roughness?: boolean;
  scale?: number;
  background?: number;
  resolution?: number | null;
}): Promise<{ q: number[]; r: (number | null)[] }> {
  return postJSON("/api/reflectivity/simulate", body);
}

/** Compute the SLD(z) depth profile for a layer stack (error-function interfaces). */
export function reflSldProfile(body: {
  layers: ReflLayer[];
  n_points?: number;
  padding?: number;
}): Promise<{ z: number[]; sld: (number | null)[] }> {
  return postJSON("/api/reflectivity/sld-profile", body);
}

// ── fit a layer model to measured data (P2.2) ───────────────────────────────

/** Request body for POST /api/reflectivity/fit (generated from the route's
 *  pydantic models in routes/reflectivity.py). */
export type ReflFitRequest = components["schemas"]["ReflFitRequest"];
export type ReflFitChannel = components["schemas"]["ReflFitChannel"];
export type ReflFitParameter = components["schemas"]["ReflFitParameter"];

/** One fitted parameter. `stderr` is null when the parameter was fixed, ended
 *  on a bound, or is not determined by the data (calc/refl_fit.py). */
export interface ReflFitParamResult {
  name: string;
  value: number;
  stderr: number | null;
  vary: boolean;
  tie: string | null;
  at_bound: boolean;
}

/** One channel's fitted points — only the points the fit used, in row order. */
export interface ReflFitCurve {
  label: string;
  spin: "+" | "-" | null;
  q: number[];
  r: number[];
  dr: number[] | null;
  model: (number | null)[];
  residual: (number | null)[];
}

/** The fit response. The route's OpenAPI response is an untyped dict, so this
 *  mirrors `fit_reflectivity`'s return value by hand. `dr` weighting reports
 *  `chi2`/`reduced_chi2`; `log` weighting reports `sum_sq_log`/
 *  `reduced_sum_sq_log` and leaves the chi-square fields null. */
export interface ReflFitResult {
  parameters: ReflFitParamResult[];
  free: string[];
  correlation: (number | null)[][];
  chi2: number | null;
  reduced_chi2: number | null;
  sum_sq_log: number | null;
  reduced_sum_sq_log: number | null;
  n_points: number;
  n_free: number;
  success: boolean;
  message: string;
  n_evaluations: number;
  weighting: "dr" | "log";
  curves: ReflFitCurve[];
  sld_profiles: { spin: "+" | "-" | null; z: number[]; sld: (number | null)[] }[];
  warnings: string[];
}

/** Fit the layer model to one or more measured curves. `signal` aborts the
 *  request; the server still finishes its own time-bounded run. */
export function reflFit(body: ReflFitRequest, signal?: AbortSignal): Promise<ReflFitResult> {
  return postJSON("/api/reflectivity/fit", body, signal);
}

// ── the posterior of a fit: DREAM through the job queue (P2.2 slice 4) ─────

export type ReflDreamRequest = components["schemas"]["ReflDreamRequest"];

/** One sampled parameter (a tied one reports its target's posterior). */
export interface ReflPosteriorParam {
  name: string;
  tie: string | null;
  median: number;
  interval68: [number, number];
  interval95: [number, number];
  /** The best draw (MAP under the flat prior). */
  map: number | null;
  rhat: number | null;
  rhat_flag: boolean;
  /** The bounds, not the data, limit the 95% interval. */
  at_bound: boolean;
}

type Percentiles = Record<"lo95" | "lo68" | "median" | "hi68" | "hi95", (number | null)[]>;

/** The finished job's result: `calc.refl_dream.sample_reflectivity`'s dict
 *  (the route's OpenAPI response is untyped, so this mirrors it by hand). */
export interface ReflPosteriorResult {
  parameters: ReflPosteriorParam[];
  free: string[];
  correlation: (number | null)[][];
  map_chi2: number | null;
  n_points: number;
  convergence: {
    converged: boolean;
    rhat_threshold: number;
    rhat_max: number | null;
    flagged: string[];
    /** Parameters whose R-hat could not be computed (too few generations). */
    unmeasured: string[];
    stopped: "completed" | "deadline" | "cancelled";
    burn: number;
    burn_requested: number;
    thin: number;
    n_chains: number;
    n_generations: number;
    n_generations_requested: number;
    n_draws: number;
    n_band_draws: number;
    n_evaluations: number;
    seed: number | null;
    reproducible: boolean;
  };
  /** R(Q) percentiles per channel, on the fitted points. */
  r_bands: ({ label: string; spin: "+" | "-" | null; q: number[]; r: number[]; dr: number[] | null } & Percentiles)[];
  /** SLD(z) percentiles per spin state, on one z grid. */
  sld_bands: ({ spin: "+" | "-" | null; z: number[] } & Percentiles)[];
  warnings: string[];
}

/** Queue a DREAM run; poll the returned job id with `pollReflJob`. A request
 *  the sampler would refuse is rejected here, before anything is queued. */
export function reflDream(
  body: ReflDreamRequest,
): Promise<{ job_id: string; plan: { n_free: number; n_chains: number; n_generations: number; n_evaluations: number } }> {
  return postJSON("/api/reflectivity/dream", body);
}

// ── the job's poll loop ──────────────────────────────────────────────────────
//
// lib/jobs.ts's `pollJob`/`cancelJob`, RESTATED here rather than imported,
// deliberately: lib/jobs lives in the lazy Curve Fit chunk, and importing it
// from this lazy chunk too splits it into a shared chunk whose name the EAGER
// entry then lists in both workshops' preload maps — measured +34 B against an
// eager budget with 51 B of headroom (2026-09-24, vite build after npm ci).
// Same transport (GET-poll /api/jobs/{id}, then /result; error text through
// http.ts's shared `ensureOk`); reflectivityJobs.test.ts runs lib/jobs as the
// oracle over the same job histories, so the two cannot drift.

/** Thrown when the job ends as cancelled — deliberate, not an error. */
export class ReflJobCancelled extends Error {
  constructor(jobId: string) {
    super(`job ${jobId} cancelled`);
    this.name = "ReflJobCancelled";
  }
}

interface JobSnap {
  status: "pending" | "running" | "done" | "error" | "cancelled";
  progress: number;
  message: string;
  error?: string;
}

/** Poll a job to its terminal state: its result on `done`; throws
 *  Error(job error) on `error` and ReflJobCancelled on `cancelled`.
 *  `onProgress` fires on every poll. */
export async function pollReflJob<T>(
  id: string,
  onProgress?: (fraction: number, message: string) => void,
  intervalMs = 1000,
): Promise<T> {
  for (;;) {
    const snap = await getJSON<JobSnap>(`/api/jobs/${id}`);
    onProgress?.(snap.progress, snap.message);
    if (snap.status === "done") return (await getJSON<{ result: T }>(`/api/jobs/${id}/result`)).result;
    if (snap.status === "error") throw new Error(snap.error || "job failed");
    if (snap.status === "cancelled") throw new ReflJobCancelled(id);
    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }
}

/** Request cooperative cancellation; the poll loop then settles as cancelled. */
export function cancelReflJob(id: string): Promise<unknown> {
  return postJSON(`/api/jobs/${id}/cancel`, {});
}
